### PR TITLE
open{jdk{8,11,17,21,23},jfx{17,21,23}}: deduplicate

### PR DESCRIPTION
**Pull request by emilazy**
Previously, the OpenJDK and OpenJFX derivations had been copied, pasted, and modified through some 14 distinct versions. This is a maintenance nightmare and resulted in a lot of inessential build configuration drift between versions. This PR aims to deduplicate them into a single expression with conditionals for the version; it should hopefully result in zero rebuilds.

The intent here is not to do anything about the madness but simply to make it legible, so as to have a single point of reference for future refactoring. I plan to do more aggressive clean‐up and add Darwin support to the derivations later, but this is what I have for 24.11. The nice thing about Nix is that as long as the resulting derivation hashes are the same, we don’t need to worry about my refactored code having significant errors.

I also rewrote the update script code. It’s in Python now, sorry; nothing against Java, but the previous state involved bootstrapping a JDK in order to run its own update script, which made adding OpenJDK 23 more of a pain than it needed to be, and I had to rework the behaviour in order to get the correct build number version handling in future. Python is the most common language for this kind of utility script in Nixpkgs after Bash, and I tried to extensively comment the code, so hopefully this won’t pose a maintenance burden for others in future.

cc @jtojnar for the update script combinator change – sorry for making it uglier.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Infinidoge** [comment]
Creating an entire Python package (and package itself) feels way overkill for an update script, which seems (hard to tell from my phone) to be contained in a single Python file. Unless there is library usage I'm not seeing, it would be preferable to condense that down to a single script file if possible.

Will make a more proper review when I'm both on my laptop and feeling better

**Comment by emilazy** [comment]
It made the development easier (since it does have external dependencies, mostly the GitHub library) and means that the linting and type‐checking can be enforced. This caught a bunch of issues while I was working on it. Since it’s used in both OpenJDK and OpenJFX, the latter of which is now in `pkgs/by-name` and therefore cannot directly access files outside of its own directory, it has to be its own package anyway.

**Comment by Infinidoge** [comment]
I'm pretty sure it's relatively straight forward to make pyproject work in a flat directory, which would be better imo. I'd rather be slightly off of Python standards than have a tiny ancillary package with an absurd amount of boilerplate structure.

Don't know how that's done off of the top of my head, but that's more to do with me not being able to think straight right now than anything else lol. Update script structure isn't a blocker by any means, and it can be cleaned up later.

**Comment by emilazy** [comment]
The way it is set up now allows doing e.g. `uv run nixpkgs-openjdk-updater …` to test it and using a standard `buildPythonApplication`. Given the existing `by-name` constraints requiring a full‐blown package, the only real boilerplate is a couple nested directories. I would prefer to stick with what Python tooling does rather than making sacrifices on the development and testing front in the absence of a strongly compelling reason to the contrary.

A bigger problem is that `nixpkgs-vet` isn’t happy with my interpolated JSON path. I might have to move things back to one big JSON file and eat the annoying merge conflicts that might result from updates.

**Comment by emilazy** [comment]
Looks like I also have to figure out what’s causing the rebuild.

**Comment by emilazy** [comment]
Okay, hopefully fixed both of those. Naturally it was of course an error in the OpenJDK 8 derivation that I accidentally fixed.

**Comment by emilazy** [comment]
Still showing a rebuild? Will look into it tomorrow.

@ofborg eval

File: pkgs/development/compilers/openjdk/generic.nix
Comment (Infinidoge) [inline pkgs/development/compilers/openjdk/generic.nix:507]: Why is this comment by itself >=17, when the actual code it refers to is >=11?

File: pkgs/development/compilers/openjdk/generic.nix
Comment (Infinidoge) [inline pkgs/development/compilers/openjdk/generic.nix:551]: Wouldn't `ls $out/lib/openjdk/bin` work?

File: pkgs/development/compilers/openjdk/generic.nix
Comment (Infinidoge) [inline pkgs/development/compilers/openjdk/generic.nix:626]: See comment near top about the double override which makes this currently redundant.

**Review by Infinidoge — CHANGES_REQUESTED** [review]
Overall looks great! Mainly some comments on code organization, plus a couple more important comments sprinkled in.

File: pkgs/common-updater/combinators.nix
Comment (emilazy) [reply pkgs/common-updater/combinators.nix:133]: As you said, it was already like this in the original code, which was not written by me. These are standard, typographically‐correct quotation marks, in normal use for hundreds of years, far longer than the straight quotes we had to use on computers for a couple decades because of memory limitations.

File: pkgs/top-level/aliases.nix
Comment (emilazy) [reply pkgs/top-level/aliases.nix:1]: These aliases-to‐throws are not cleaned up by `maintainers/scripts/remove-old-aliases.py`, which caused manual toil related to evaluation issues recently when cleaning up old `throw`s. We should avoid this pattern due to the automation issues it causes.

File: pkgs/development/compilers/openjdk/generic.nix
Comment (emilazy) [reply pkgs/development/compilers/openjdk/generic.nix:126]: Ah, that is why the rebuilds weren’t fixed: I forgot to remove the override from the header. Fixed.

File: pkgs/development/compilers/openjdk/generic.nix
Comment (emilazy) [reply pkgs/development/compilers/openjdk/generic.nix:130]: Lazy evaluation means that the `assert` is the first thing to be evaluated:

```
emily@yuyuko ~/D/nixpkgs ((afb8cd5d))> nix repl -f .
Nix 2.24.9
Type :? for help.
Loading installable ''...
Added 22519 variables.
nix-repl> :b callPackage ./pkgs/development/compilers/openjdk/generic.nix { featureVersion = "123"; } 
error:
       … while calling a functor (an attribute set with a '__functor' attribute)
         at /Users/emily/Developer/nixpkgs/lib/customisation.nix:264:13:
          263|     in if missingArgs == {}
          264|        then makeOverridable f allArgs
             |             ^
          265|        # This needs to be an abort so it can't be caught with `builtins.tryEval`,

       … while evaluating a branch condition
         at /Users/emily/Developer/nixpkgs/lib/customisation.nix:148:7:
          147|     in
          148|       if isAttrs result then
             |       ^
          149|         result // {

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: OpenJDK 123 is not a supported version
```

File: pkgs/development/compilers/openjdk/generic.nix
Comment (emilazy) [reply pkgs/development/compilers/openjdk/generic.nix:256]: No, the ordering matters for the derivation hash. This is not meant to be a clean‐up but rather an accurate 0‐rebuild refactoring of the existing derivation, so that we can be sure the refactoring is correct. I will clean it up separately.

File: pkgs/development/compilers/openjdk/generic.nix
Comment (emilazy) [reply pkgs/development/compilers/openjdk/generic.nix:313]: They do, because these lists factor in to the final `.drv` file. That’s why I’ve marked the duplicates and maintained the version‐conditional order.

Comment (emilazy) [reply pkgs/development/compilers/openjdk/generic.nix:313]: BTW, the ordering is semantically meaningful when multiple packages contain the same libraries or similar.

File: pkgs/development/compilers/openjdk/generic.nix
Comment (emilazy) [reply pkgs/development/compilers/openjdk/generic.nix:407]: Flag lists are ordered. However, most of the complexity here is not related to ordering; I believe only the devendoring flags are due to that. I plan to do a major clean‐up of this derivation for the 25.05 cycle and will rationalize this then.

File: pkgs/development/compilers/openjdk/generic.nix
Comment (emilazy) [reply pkgs/development/compilers/openjdk/generic.nix:503]: Because it was changed between `11.nix` and `17.nix`, and this affects the derivation hash.

File: pkgs/development/compilers/openjdk/generic.nix
Comment (emilazy) [reply pkgs/development/compilers/openjdk/generic.nix:547]: That would change the derivation hash.

File: pkgs/development/compilers/openjdk/generic.nix
Comment (emilazy) [reply pkgs/development/compilers/openjdk/generic.nix:622]: I changed it to be like this because the OpenJDK 8 derivation incorrectly lists the non‐overrided bootstrap JDK here, making it quite useless, but affecting the derivation hash. I’ve removed the override from the function header for now so that this can correctly match the original derivation and avoid the rebuilds.

File: pkgs/common-updater/combinators.nix
Comment (jtojnar) [inline pkgs/common-updater/combinators.nix:125]:
```suggestion
        if hasCommitSupport then
          # Exactly one update script declares only “commit” feature and all the rest declare only “silent” feature.
```

File: pkgs/common-updater/combinators.nix
Comment (jtojnar) [inline pkgs/common-updater/combinators.nix:127]:
```suggestion
        else if hasSilentSupport then
          # All update scripts declare only “silent” feature.
```

File: pkgs/common-updater/combinators.nix
Comment (jtojnar) [inline pkgs/common-updater/combinators.nix:129]:
```suggestion
        else
          # No update script declares any supported feature to fail loudly on unknown features rather than silently discard them.
```

File: pkgs/common-updater/combinators.nix
Comment (jtojnar) [inline pkgs/common-updater/combinators.nix:124]:
```suggestion
      # Supported features currently only describe the format of the standard output of the update script.
      # Here we ensure that the standard output of the combined update script is well formed.
      validateFeatures =
```

**Review by jtojnar — COMMENTED** [review]
The update script changes look good to me.

Unfortunately, the logic does not translate to Nix 1:1 so we can probably only make it clearer by introducing comments.

**Comment by emilazy** [comment]
Good idea, I’ve applied the comments. I was working out a clearer approach with less code duplication based on a fold, but then I realized that it’s not clear that the current behaviour is correct in all cases anyway, e.g. if you combine a `commit` update script that explicitly enumerates the changed files with a `silent` update script that expects the updater to infer the additional files it changes. Update scripts are a bit of a badly‐composing mess (see e.g. my inability to just reuse `finalAttrs.mitmCache.updateScript` in OpenJFX), I’m thinking about trying to rationalize the system for 25.05… (But thank you for adding these neat combinators!)

**Comment by emilazy** [comment]
Base branch eval error was fixed.

@ofborg eval

File: pkgs/by-name/ni/nixpkgs-openjdk-updater/package.nix
Comment (thiagokokada) [inline pkgs/by-name/ni/nixpkgs-openjdk-updater/package.nix:1]: Small nit, but maybe we should call this `_nixpkgs-openjdk-updater` since this is not supposed to be exported in all-packages and I don't want to go through the aliases shenanigans if we ever want to remove this script.

But not strong opinions here, if you think this is fine @emilazy it is to me.

Comment (emilazy) [reply pkgs/by-name/ni/nixpkgs-openjdk-updater/package.nix:1]: I viewed this as similar to things like `vimPluginsUpdater` and `nix-update-script`: helpers oriented towards Nixpkgs use but not technically completely coupled to it. I feel the export is a good thing for testing and personally feel like the current name is fine. I think it would be fine to just add a simple `throw` if this is removed since `nixpkgs-*` makes it clear who the target audience is, but I also don’t really care about what the name is and don’t want to be difficult so if you’d like an underscore in front of it I can do that. It seems like currently the underscore prefix is only used in `by-name` for packages starting with a number.

**Review by thiagokokada — APPROVED** [review]

**Comment by emilazy** [comment]
Investigating the handful of remaining rebuilds.

File: pkgs/top-level/aliases.nix
Comment (Infinidoge) [reply pkgs/top-level/aliases.nix:1]: Yep that makes sense. Definitely a good idea then.

File: pkgs/development/compilers/openjdk/generic.nix
Comment (Infinidoge) [reply pkgs/development/compilers/openjdk/generic.nix:256]: I have less faith in Nix today than I did yesterday. Input packages are seriously not sorted and deduplicated?? Gah.

Comment (thiagokokada) [reply pkgs/development/compilers/openjdk/generic.nix:256]: I think this is the difference of using a list instead of a set (not that Nix has a native set data type), and like @emilazy said this actually can cause a build to differ (e.g.: this will change the order of applications appears in `$PATH` for `nativeBuildInputs`), so I don't think this is unresonable behavior.

File: pkgs/development/compilers/openjdk/generic.nix
Comment (Infinidoge) [reply pkgs/development/compilers/openjdk/generic.nix:313]: There can definitely be semantic meaning, but it feels messier with it as split up as it is. Given it causes rebuilds this can be marked resolved and thought about later.

**Comment by emilazy** [comment]
The rebuilds are from the fact that I do `openjfx ? …default…` to support the existing pattern of overriding `openjfx`, but of course `openjfx` is an existing alias and that results in the default never being used. Thinking about what to do here to hopefully maintain interface compatibility without compromising the ability to move OpenJDK into `by-name`. It may be that the best option is to break `by-name` compatibility for now and batch interface‐breaking changes when doing further refactors. Or just fix the few in‐tree users and accept a minor breaking change to the interface; there’s already technically some slight differences (e.g. the `openjdk21-bootstrap` argument is gone). Open to thoughts here.

**Review by Infinidoge — APPROVED** [review]
Aside from the various things that will need to be touched on a rebuild-allowed followup, looks great to me! Thanks for putting in the heavy legwork in getting this refactor done!

**Comment by thiagokokada** [comment]
> The rebuilds are from the fact that I do `openjfx ? …default…` to support the existing pattern of overriding `openjfx`, but of course `openjfx` is an existing alias and that results in the default never being used. Thinking about what to do here to hopefully maintain interface compatibility without compromising the ability to move OpenJDK into `by-name`. It may be that the best option is to break `by-name` compatibility for now and batch interface‐breaking changes when doing further refactors. Or just fix the few in‐tree users and accept a minor breaking change to the interface; there’s already technically some slight differences (e.g. the `openjdk21-bootstrap` argument is gone). Open to thoughts here.

I am fine with the rebuilds here as long it affects just a small number of packages. 11-100 seems reasonable for a big PR as this one.

File: pkgs/development/compilers/openjdk/generic.nix
Comment (emilazy) [reply pkgs/development/compilers/openjdk/generic.nix:313]: Yeah, I will definitely try to clean this stuff up for 25.05. It will need adjusting for Darwin anyway.

**Comment by emilazy** [comment]
Well, the rebuilds are not benign: they are the canary in the coal mine telling us that overriding `openjfx` is broken. Packages in the tree rely on that. That’s why I need to choose between the potential fixes I listed.

**Comment by thiagokokada** [comment]
> Well, the rebuilds are not benign: they are the canary in the coal mine telling us that overriding `openjfx` is broken. Packages in the tree rely on that. That’s why I need to choose between the potential fixes I listed.

Between the 2 options, I think the "Or just fix the few in‐tree users and accept a minor breaking change to the interface" is the most reasonable one.

**Comment by emilazy** [comment]
I’ve renamed the parameter to `openjfx_jdk`, i.e. “the version of OpenJFX to use with this JDK”, following the naming convention for multi‐version packages and preferring `jdk` over `openjdk` like `jdk-bootstrap` in the vague hopes that there could be uniformity between different possible JDKs in the interface here. I don’t really like that parameter name at all, but I didn’t get enough sleep to think of a better one, the deadline for breaking changes is in about 10 hours, and it’s likely there will be other breaking changes in 25.05 as I do a larger clean‐up, so I think bikeshedding can wait until then unless anyone *really* hates the name.

This should be ready to merge once ofborg says no rebuilds.

File: pkgs/development/compilers/openjdk/generic.nix
Comment (skeuchel) [inline pkgs/development/compilers/openjdk/generic.nix:654]:
```suggestion
      ++ lib.optionals atLeast17 [
        "riscv64-linux"
      ];
```

This would make https://github.com/NixOS/nixpkgs/pull/343907 obsolete.

Comment (emilazy) [reply pkgs/development/compilers/openjdk/generic.nix:654]: Would you be up for rebasing your PR after this one merges? I’d be happy to review and merge it then (though I don’t have hardware to test it out). I’ve tried to keep this PR strictly to changes that result in no change in the resulting derivations and `meta`s here in order to separate *organizing the current state of things* from *deciding what the state of things ought to be*; keeping it purely mechanical means that this PR can’t be responsible for any regressions, should never need to be reverted, and doesn’t express any policy other than “copy‐pasted code is bad”. Not that I expect adding RISC‐V support would lead to any of those, which is why I’m happy to hit the merge button after :)

Comment (skeuchel) [reply pkgs/development/compilers/openjdk/generic.nix:654]: Sure, no problem.

**Comment by emilazy** [comment]
One rebuild for the updater package 🎉 🎉 🎉

This is ready now.

**Review by thiagokokada — APPROVED** [review]

**Comment by thiagokokada** [comment]
Thanks for the great work again @emilazy.

**Comment by jlesquembre** [comment]
Thanks for working on this, but I don't see the advantage of rewriting the update script.

What's the painful part of running Java in the update script? The main reason for using Java was to avoid re-implementing the version parse logic. (https://github.com/NixOS/nixpkgs/pull/306616#discussion_r1615258096). Also, it looks like the Python version doesn't work as expected: https://nixpkgs-update-logs.nix-community.org/jdk23/2024-10-31.log
It updates the version to `refs/tags/jdk-23.0.1+11` (BTW, why not `jdk-23.0.1-ga`, I find the `-ga` suffix more user friendly. In my opinion, the build version number should only be used for non-GA versions)

**Comment by emilazy** [comment]
`refs/tags/jdk-23.0.1+11` is a correct version to update to, what’s the problem? We deliberately pick the tags corresponding to GA releases but without the `-ga` suffix so that we have the build number information available. That will allow us to give proper full `java -evrsion` output matching other Java builds like Temurin and Zulu, which identify their builds with the full version including build number and pass that build number information down to the OpenJDK build process.

The problem with the Java dependency is that having to build an OpenJDK version to update that same OpenJDK version causes problems when adding new versions and causes huge builds when doing it on top of `staging`. (It also made it basically impossible for me to run on macOS, where our OpenJDK packaging still needs work.) The logic needed changing to include the build number and to handle OpenJFX repositories and I was more comfortable doing it in Python than Java, and @Infinidoge, the other maintainer who has been working on cleaning up our OpenJDK packaging, agreed that it was preferable to have it in that language. Nixpkgs maintainers that aren’t familiar with a given language ecosystem will sometimes have to use and modify these scripts, and Python is the most common language after Bash for tooling tasks like these in the tree. Since the version format is documented as a simple regular expression by OpenJDK, and the modifications to handle the OpenJFX repositories meant that it needed modifying anyway, I don’t think the duplication is a problem. I am not even sure `java.lang.Runtime.Version` handles the JDK 8 version format, though I admittedly did not check.

**Comment by jlesquembre** [comment]
> the other maintainer who has been working on cleaning up our OpenJDK packaging, agreed that it was preferable to have it in that language 
> It also made it basically impossible for me to run on macOS

Fair enough, I didn't consider macOS (I don't have a mac, but it'd be nice to improve the macOS build)

> refs/tags/jdk-23.0.1+11 is a correct version to update to, what’s the problem?

In that case, something else is going on, or I'm missing something. The update script was executed, but I don't see a PR to update jdk23 to `23.0.1`

**Comment by emilazy** [comment]
Yeah, I’d like to get our OpenJDK and OpenJFX packages building on Darwin so that we can use a source‐built JDK on all platforms and probably drop Zulu as redundant.

I already did the update in https://github.com/NixOS/nixpkgs/pull/351025, which might be it. It’s possible there’s something else going on too, though, since r-ryantm can be a bit finnicky. If we don’t get the next batch of updates automatically I’ll look into it.

**Pull request merged**
